x/reports/internal/keeper/keeper.go

```diff
@@ -0,0 +1,100 @@
+package keeper
+
+import (
+	"bytes"
+
+	"github.com/cosmos/cosmos-sdk/codec"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/desmos-labs/desmos/x/posts"
+	"github.com/desmos-labs/desmos/x/reports/internal/types"
+	"github.com/desmos-labs/desmos/x/reports/internal/types/models"
+)
+
+// Keeper maintains the link to data storage and exposes getter/setter methods for the various parts of the state machine
+type Keeper struct {
+	PostKeeper posts.Keeper // post's keeper to perform checks on the postIDs
+	StoreKey   sdk.StoreKey // Unexposed key to access store from sdk.Context
+	Cdc        *codec.Codec // The wire codec for binary encoding/decoding.
+}
+
+// NewKeeper creates new instances of the reports Keeper
+func NewKeeper(pk posts.Keeper, cdc *codec.Codec, storeKey sdk.StoreKey) Keeper {
+	return Keeper{
+		PostKeeper: pk,
+		StoreKey:   storeKey,
+		Cdc:        cdc,
+	}
+}
+
+// SaveReport allows to save the given reports inside the current context.
+// It assumes that the given reports has already been validated.
+// If the same reports has already been inserted, nothing will be changed.
+func (k Keeper) SaveReport(ctx sdk.Context, postID posts.PostID, report types.Report) bool {
+	store := ctx.KVStore(k.StoreKey)
+	key := models.ReportStoreKey(postID)
+	// Get the list of reports related to the given postID
+	var reports models.Reports
+	k.Cdc.MustUnmarshalBinaryBare(store.Get(key), &reports)
+
+	// try to append the given reports
+	reports, appended := reports.AppendIfMissing(report)
```

Comment:
i'm thinking that maybe this way of checking if the report already exists could lead to a problem if there are a lot of reports for the same post (which can easily happen if it contains nudity for example). 

Also, I don't see why we would want to block multiple reports to the same post (even from the same user). Instead, what I think we should do is allow a user to report as many times he wants the same post. It will be the dApp responsibility to handle that.

Reply:
I was thinking that from the "spam-side" but I keep forgetting that they will spend token each report they'll made so... 🤣